It seems there is a failing test:

```
app = <SphinxTestApp buildername='html'>

    @pytest.mark.sphinx('html', testroot='ext-autodoc')
    def test_autodoc_TYPE_CHECKING(app):
        options = {"members": None,
                   "undoc-members": None}
        actual = do_autodoc(app, 'module', 'target.TYPE_CHECKING', options)
>       assert list(actual) == [
            '',
            '.. py:module:: target.TYPE_CHECKING',
            '',
            '',
            '.. py:class:: Foo()',
            '   :module: target.TYPE_CHECKING',
            '',
            '',
            '   .. py:attribute:: Foo.attr1',
            '      :module: target.TYPE_CHECKING',
            '      :type: ~_io.StringIO',
            '',
            '',
            '.. py:function:: spam(ham: ~collections.abc.Iterable[str]) -> tuple[gettext.NullTranslations, bool]',
            '   :module: target.TYPE_CHECKING',
            '',
        ]
E       AssertionError: assert ['', '.. py:m...HECKING', ...] == ['', '.. py:m...HECKING', ...]
E         At index 13 diff: '.. py:function:: spam(ham: ~collections.abc.Iterable[str]) -> tuple[~gettext.NullTranslations, bool]' != '.. py:function:: spam(ham: ~collections.abc.Iterable[str]) -> tuple[gettext.NullTranslations, bool]'
E         Use -v to get more diff
```

I don't know if this is the intended behavior though. After all the result of `stringify_annotation("gettext.NullTranslations")` is actually `gettext.NullTranslations`, not `~gettext.NullTranslations`.
Is there anybody that can tell me whether the test case is legit or not?

--- a/sphinx/util/typing.py
+++ b/sphinx/util/typing.py
@@ -276,11 +276,14 @@ def stringify_annotation(
     elif str(annotation).startswith('typing.Annotated'):  # for py310+
         pass
     elif annotation_module == 'builtins' and annotation_qualname:
-        if (args := getattr(annotation, '__args__', None)):  # PEP 585 generic
-            return (
-                f'{annotation_qualname}['
-                f'{", ".join(stringify_annotation(arg, mode=mode) for arg in args)}]'
-            )
+        if (args := getattr(annotation, '__args__', None)) is not None:  # PEP 585 generic
+            if not args:
+                return repr(annotation)
+            else:
+                return (
+                    f'{annotation_qualname}['
+                    f'{", ".join(stringify_annotation(arg, mode=mode) for arg in args)}]'
+                )
         else:
             return annotation_qualname
     elif annotation is Ellipsis:

--- a/sphinx/util/typing.py
+++ b/sphinx/util/typing.py
@@ -277,7 +277,10 @@ def stringify_annotation(
         pass
     elif annotation_module == 'builtins' and annotation_qualname:
         if (args := getattr(annotation, '__args__', None)):  # PEP 585 generic
-            return f'{annotation_qualname}[{", ".join(stringify_annotation(arg) for arg in args)}]'
+            return (
+                f'{annotation_qualname}['
+                f'{", ".join(stringify_annotation(arg, mode=mode) for arg in args)}]'
+            )
         else:
             return annotation_qualname
     elif annotation is Ellipsis:

--- a/sphinx/util/typing.py
+++ b/sphinx/util/typing.py
@@ -276,8 +276,8 @@ def stringify_annotation(
     elif str(annotation).startswith('typing.Annotated'):  # for py310+
         pass
     elif annotation_module == 'builtins' and annotation_qualname:
-        if hasattr(annotation, '__args__'):  # PEP 585 generic
-            return repr(annotation)
+        if (args := getattr(annotation, '__args__', None)):  # PEP 585 generic
+            return f'{annotation_qualname}[{", ".join(stringify_annotation(arg) for arg in args)}]'
         else:
             return annotation_qualname
     elif annotation is Ellipsis:

